Adjoint variables through diagnostics package

timothyas opened the pull request:
## What changes does this PR introduce?
This enables the user to print the adjoint variables: adTheta, adSalt, adUvel, adVvel, adWvel through the diagnostics package, allowing one to compute averaged quantities etc. as with other diagnostic variables. adEtaN is included as these variables are, but it is currently broken in `addummy_in_stepping.F`, printing zeros at each time step. 

Brief documentation is given. 

## What is the current behaviour? 
Currently, users can only print snapshots of all adjoint variables at a frequency of `adjDumpFreq`.


## What is the new behaviour 
Now users can select specific variables for printing, and can compute averaged quantities just as with other diagnostic variables. The old 'adjoint dump' method is not removed and can be used as well.

## Does this PR introduce a breaking change? 
No, although adEtaN is broken in `addummy_in_stepping.F`, and this does not fix that problem.

## Other information:
- This is only available for MDS I/O.
- OpenAD has not been tested, but _should_ work fine
- The statistical computations in the diagStats routines of the diagnostics package are not available. 
- Not all adjoint variables in `addummy_in_stepping.F` are included, only the main ocean state variables.

If there is interest, it would not be difficult to include more variables or enable NetCDF I/O.

edoddridge commented on doc/outp_pkgs/outp_pkgs.rst:
```diff
@@ -383,6 +383,75 @@ of separate output streams that the user specifies in the namelist
+
+Note: the diagnostics package automatically provides a phase shift of :math:`frequency/2`, 
+so specify `frequency = 0` to match output from `adjDumpFreq`.
+
```
Presumably this should be "specify `timePhase = 0`", rather than frequency?

timothyas replied:
Yep, thanks for the catch. Fixed.

heimbach commented:
Hi @edoddridge,

let's keep it simple for now. 
Tim Smith has made a tremendous effort at making adjoint output available via diagnostics package. This is a qualitative improvement. 

I don't want him to deal with the NetCDF extension for now. 
If you like, put a STOP statement in case someone want this output in NetCDF.

Thanks
-Patrick

> On Jun 11, 2018, at 9:52 AM, Ed Doddridge <notifications@github.com> wrote:
> 
> @edoddridge commented on this pull request.
> 
> What's the barrier to adding support for NetCDF? Once the variables are sent to the diagnostics package both options should be avaiable?
> 
> In doc/outp_pkgs/outp_pkgs.rst:
> 
> > +      fields(1:5,1) = 'ADJtheta','ADJsalt ',
> +                         'ADJuvel ','ADJvvel ','ADJwvel '
> +      filename(1) = 'diags/adjState_3d_snaps',
> +      frequency(1)=-86400.0,
> +      timePhase(1)=0.0,
> +    #---
> +      fields(1:5,2) = 'ADJtheta','ADJsalt ',
> +                         'ADJuvel ','ADJvvel ','ADJwvel '
> +      filename(2) = 'diags/adjState_3d_avg',
> +      frequency(2)= 86400.0,
> +    #---
> +    &
> +
> +Note: the diagnostics package automatically provides a phase shift of :math:`frequency/2`, 
> +so specify `frequency = 0` to match output from `adjDumpFreq`.
> +
> 
> Presumably this should be "specify timePhase = 0", rather than frequency?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 


--------------------------------------------------------
Patrick Heimbach, Ph.D. | http://heimbach.wordpress.com

| The University of Texas at Austin, ICES/JSG/UTIG |
201 East 24th Street, POB 4.232 | Austin, TX 78712 | USA
FON: +1-512-232-7694 | Email: heimbach@utexas.edu

| Massachusetts Institute of Technology, EAPS/PAOC |
77 Massachusetts Ave, 54-1420 | Cambridge MA 02139 | USA
FON: +1-617-253-5259 | Email: heimbach@mit.edu
--------------------------------------------------------

christophernhill commented:
@edoddridge I think that it is hard to produce netCDF in reverse mode at present. Some of the netcdf related functions are not shown to the differentiation tools TAF and OpenAD. They are hidden so that the dataflow analysis doesn't get confused. Fixing that might be good long term issue, but maybe not for this PR.

timothyas commented:
Hey all, 
As @heimbach mentioned, I was previously unfamiliar with netCDF so I just wanted to get the basics working with the familiar MDS framework first. However, it turned out netCDF output was almost there with the original PR. I was able to add one tiny if statement to `diagnostics_out.F` and now the netCDF output works as well. 

I tested this with a small setup, and both forward (regular) and adjoint diagnostic output is identical between MDS and netCDF. Documentation has been updated, removing the comment about netCDF not working.

gaelforget commented:
I agree with others that netcdf is low priority here. @edoddridge it would be great if you could double check that hand-coded adjoint pieces are correct and work as expected before this gets pushed (we lack an automated mechanism to detect when things like addummy_in_stepping are incorrect or break)

timothyas commented:
@gaelforget out of the variables I included in this PR, adEtaN is still broken - it prints zeros at all time steps from `addummy_in_stepping`. My original thought was that this PR would address printing through the diagnostics package on its own, and fixing adEtaN (and possibly other variables, if they are broken and are to be added) would be a separate issue. But of course, this is up to you guys. 

Also, I was able to get netCDF working with the latest commit.

edoddridge commented:
Nice work @timothyas! I naively assumed it would be fairly straightforward since you'd already done the hard work of getting `pkg/diagnostics` to play nicely. Glad that turned out to be the case.

My vote is to keep this PR focussed on using `pkg/diagnostics` and to leave the problems with `adEtaN` for a future PR. Happy to be overruled by the adjoint folks though.

@gaelforget - I'm definitely not the right person to double check the adjoint code. I've actually never run it. Looks like @jm-c has taken this PR though, so I expect he'll do a thorough review of the code before anything gets merged.

jm-c commented:
@timothyas, few comments:
1) there are significant changes/modifications to pkg/diagnostics so I think
 it's good for now that they are within #ifdef DIAGNOSTICS_ADJ_OUTPUT ;
 this new CPP option should be added to pkg/diagnostics/DIAG_OPTIONS.h,
 and the default, for now, could be #undef I guess.
2) I would like to keep diagnostics short names (that are listed in data.diagnostics
  and within the code) free of any special meaning and code treatment.
  Instead of testing for flds()(1:3) = 'ADJ', there is a 10 character code (gdiag/diagCode)
  that can be use to recognize an adjoint variable diagnostics from a normal/forward one.
  I could suggest to use the 4th position (currently only used to document
  "positive definite" variable) with something like diagCode(4:4) = 'A' ?
3) this implementation does not work when mixing adjoint and forward diagnostics
  in the same output file. I tried, in data.diagnostics:
> fields(1:2,4)  = 'THETA   ','ADJtheta',
>   fileName(4) = 't_Diag',
>  frequency(4) = 172800.,
  but this does not output ADJtheta correctly (I got warnings); I also tried:
> fields(1:2,5)  = 'ADJsalt ','SALT    ',
  and this does not give warning but "SALT" output is strange.
  I agree that this would be difficult to fix but a) it needs to be documented,
  b) should have some check & stop if there is a mix of ADJ & FWD diag in any 
  diagnostics stream (also if DIAGNOSTICS_ADJ_OUTPUT is undef).
4) I also tried to run with one Stats-Diag output file (but only with forward
  diagnostics and no ADJ diags). This is not great because I got plenty
  of warning (diagnostics has not been filled ...) for each diagnostics in DIAG_STATIS_PARMS 
  and each output iteration ; and it did created a "fort.9" output file after
  my normal stats-diag output file was closed.
5) given point (3) above, I think that most of the modifications in diagnostics_out.F
  (test if forward diag or adj diag ...) could be removed and replaced with fewer ones
  in diagnostics_write.F, to only call S/R DIAGNOSTICS_OUT when needed.
  Might be worth also to create a new version of diagnostics_write.F (diagnostics_adj_write ?)
  just for this purpose (called from addummy_in_stepping.F).

timothyas commented:
Thanks for the detailed comments @jm-c. Re: 

1. Done with the latest commit. 

2. These flags are totally unfamiliar to me, so I took your suggestion and made the test gcode(4:4)=‘A’ rather than flds(1,n)(1:3)=‘ADJ’

3 & 5. I agree with cleaning up `diagnostics_write.F`, so I wrote `diagnostics_write_adj.F` to do the output just for these variables and stripped `diagnostics_write.F` of everything except to make sure the diagnostic variables are not adjoint variables. Also, the issue of mixing adj/fwd variables has been documented and has a check & stop in `diagnostics_check.F`.

Just to clarify, do you agree with my snapshot time step convention: that myIter is kept at myIter, not myIter-1 for adjoint snapshots? This is to replicate the old, adjDump output (and I think is more intuitive). 

4. I'm not sure how the errors were being printed in diag stats. I am hoping this got fixed by doing adjoint writing in a separate file, but if not it might have to do with the fact that the package doesn't get turned off in adjoint mode (at model end). I can look into this next week.

jm-c commented:
@timothyas Thanks for making these changes. Seems that you though carefully about snapshot time step
issue so I would follow what you coded. I am still checking how it works, might push directly some changes
to your PR and from there we will see if every one is happy with this new feature.

jm-c commented:
@timothyas  I just made some changes that fix some problems I found (see commit msg for addummy_in_stepping.F). It does change the output (because of previously missing some AdExch) so if you could take a look and check that would be good. 
Other points:
1) I am going to see if we could remove completely DIAGNOSTICS_ADJ_OUTPUT and use
 instead the new internal parameter "useDiag4AdjOutp".
2) I have the impression that we can now use the unmodified "diagnostics_out.F"
  and it works just fine. Could you check ?
3) I tried OpenAD but there are other problems (unrelated to Adj Diags) with some S/R in pkg/diagnostics
 that do not compile with OpenAD. So for now it's untested.
4) Adj Diags does not work with divided adjoint; Might need to revisit when and from where
TURNOFF_MODEL_IO is called. I suggest to leave this DivAdj issue on the side for now and 
address this in separated PR.

jm-c commented:
@timothyas I put an alternative implementation (in my own git-repos, branch "adj_var_diagnostics") that does not use CPP option DIAGNOSTICS_ADJ_OUTPUT but use instead this new internal parameter "useDiag4AdjOutp".
This might be more convenient (no need to have this option defined in local DIAG_OPTIONS.h), making
adj-diags feature available as soon as pkg/diagnosctics is compiled. Could be more efficient too (no 
additional ad-exch) in case none of the adj-diags are used (in this case useDiag4AdjOutp=F). If you want to give it a try when you have time. 
Also I don't think we need this new parameter "writeDiagnosticsAdj" so that we can keep the original
AUTODIFF_PARAMS.h, autodiff_inadmode_set/unset_ad.F, autodiff_readparms.F and diagnostics_out.F

christophernhill commented:
@timothyas did you get a chance to look at @jm-c comments? I think he is waiting to hear back when you have a chance. 

Also there is a lurking error in the doc 
 ( /home/travis/build/MITgcm/MITgcm/doc/outp_pkgs/outp_pkgs.rst:207:Malformed table. )
- no good deed goes unpunished by Travis!

timothyas commented:
Hey @jm-c & @christophernhill - so sorry for the delay. I was traveling and so was not quite able to give these changes a whirl until today. Thanks for the feedback! 

@jm-c - 
Noted and thanks for the fixes to `addummy_in_stepping.F`

1. I checked out your forked repo/branch and this works perfectly! I actually like this solution better than my original because there is no compile time option that needs to be hassled with. I'm not sure how to properly merge your branch with this PR, but my vote would be to use that setup. 
2. I checked and this does work with the old `diagnostics_out.F`. I reverted this file back in commit 
4f1c62c. 
3&4. Noted and this is mentioned in the documentation (commit 
9546744). 

Let me know if you'd also like a brief line in doc/tag-index about this PR as well.

@christophernhill - 
Thanks for the catch, I had been ignoring travis because it seemed like all manual builds were failing and somehow missed the warning when building locally.

jm-c commented:
@timothyas here is what I propose: I am going to try to merge my branch into your PR ; this way, we keep you as author. Once it's done, you could make 2 additional changes: 1) short add to doc/tag-index and 2) update outp_pkgs.rst by removing DIAGNOSTICS_ADJ_OUTPUT there. Also, would be nice to mention that ALLOW_AUTODIFF_MONITOR needs to be #define (either in AUTODIFF_OPTIONS.h or in ECCO_CPPOPTIONS.h if this one is used). And after we could merge (squash and merge) this PR.
Let me know if this works for you.

timothyas commented:
@jm-c that sounds great. I'll await your merge and make those changes shortly after.

jm-c commented:
@timothyas I finished to bring in files from my development branch. I also change one AD verification
experiment to write out some adj-diag output files. I would suggest to update this branch with latest
main repos changes before updating tag-index, so that we don't get a merge conflict. I don't think there are so many new changes (since you already have all SIZE.h changes) so you could even follow the manual (bottom of section 5.2.1) instructions ; or, if you prefer, you can rebase.

timothyas commented:
Thanks @jm-c. I added those notes and also mentioned your verification example to the documentation. I rebased the branch to keep the commit history clean. Let me know if you want any additions to this PR.
